File: urbanairship-core/src/main/java/com/urbanairship/push/AirshipNotificationManager.java
/* Copyright Airship and Contributors */

package com.urbanairship.push;

import android.content.Context;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.core.app.NotificationManagerCompat;

/**
 * Airship notification manager wrapper.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public interface AirshipNotificationManager {

    enum PromptSupport {
        NOT_SUPPORTED,
        COMPAT,
        SUPPORTED;
    }

    boolean areNotificationsEnabled();

    boolean areChannelsCreated();

    @NonNull
    PromptSupport getPromptSupport();

    static AirshipNotificationManager from(@NonNull Context context) {
        NotificationManagerCompat notificationManagerCompat = NotificationManagerCompat.from(context);
        int targetSdkVersion = context.getApplicationInfo().targetSdkVersion;

        return new AirshipNotificationManager() {
            @Override
            public boolean areNotificationsEnabled() {
                return notificationManagerCompat.areNotificationsEnabled();
            }

            @Override
            public boolean areChannelsCreated() {
                return !notificationManagerCompat.getNotificationChannelsCompat().isEmpty();
            }

            @NonNull
            @Override
            public PromptSupport getPromptSupport() {
                if (Build.VERSION.SDK_INT >= 33) {
                    if (targetSdkVersion >= 33) {
                        return PromptSupport.SUPPORTED;
                    } else {
                        return PromptSupport.COMPAT;
                    }
                } else {
                    return PromptSupport.NOT_SUPPORTED;
                }
            }
        };
    }

}

File: urbanairship-core/src/main/java/com/urbanairship/push/AirshipNotificationManager.kt
/* Copyright Airship and Contributors */
package com.urbanairship.push

import android.content.Context
import android.os.Build
import androidx.annotation.RestrictTo
import androidx.core.app.NotificationManagerCompat

/**
 * Airship notification manager wrapper.
 *
 * @hide
 */
internal interface AirshipNotificationManager {

    enum class PromptSupport { NOT_SUPPORTED, COMPAT, SUPPORTED }

    fun areNotificationsEnabled(): Boolean

    fun areChannelsCreated(): Boolean

    val promptSupport: PromptSupport

    companion object {

        fun from(context: Context): AirshipNotificationManager {
            val notificationManagerCompat = NotificationManagerCompat.from(context)
            val targetSdkVersion = context.applicationInfo.targetSdkVersion

            return object : AirshipNotificationManager {
                override fun areNotificationsEnabled(): Boolean {
                    return notificationManagerCompat.areNotificationsEnabled()
                }

                override fun areChannelsCreated(): Boolean {
                    return notificationManagerCompat.notificationChannelsCompat.isNotEmpty()
                }

                override val promptSupport: PromptSupport
                    get() {
                        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                            if (targetSdkVersion >= Build.VERSION_CODES.TIRAMISU) {
                                PromptSupport.SUPPORTED
                            } else {
                                PromptSupport.COMPAT
                            }
                        } else {
                            PromptSupport.NOT_SUPPORTED
                        }
                    }
            }
        }
    }
}
